<div class="alert alert-success">
<b>Reviewer's comment V2</b>

The project is accepted! Thanks for taking the time to improve it! I left a couple of new comments below to clarify some thins, please check them out! And good luck on the next sprint!

</div>

**Review**

Hi, my name is Dmitry and I will be reviewing your project.
  
You can find my comments in colored markdown cells:
  
<div class="alert alert-success">
  If everything is done successfully.
</div>
  
<div class="alert alert-warning">
  If I have some (optional) suggestions, or questions to think about, or general comments.
</div>
  
<div class="alert alert-danger">
  If a section requires some corrections. Work can't be accepted with red comments.
</div>
  
Please don't remove my comments, as it will make further review iterations much harder for me.
  
Feel free to reply to my comments or ask questions using the following template:
  
<div class="alert alert-info">
  For your comments and questions.
</div>
  
First of all, thank you for turning in the project! You did a great job overall! There are only a couple of small issues that need to be fixed before the project is accepted. Let me know if you have any questions!

# Beta Bank Churn Model

# Contents <a id='back'></a>

* [Introduction](#introduction)
* [Data Overview](#data_overview)
    * [Initialization](#initialization)
    * [Load Data](load_data)
* [Prepare the Data](#prepare_data)
    * [Check for Duplicates](#duplicates)
    * [Check for Missing Values](#missing_values)
    * [Converting Data Types](#data_types)
* [Class Balance Examination](#class_balance)
    * [Model without Accounting for Imbalance](#raw_model)
    * [Fixing Class Imbalance](#fixing_class_imbalance)
        * [Upsampling Random Forest](#upsampling_random_forest)
        * [Downsampling Random Forest](#downsampling_random_forest)
        * [Upsampling Decision Tree](#upsampling_decision_tree)
        * [Downsampling Decision Tree](#downsampling_decision_tree)
    * [Final Model](#final_model)
* [Conclusion](#conclusion)

# Introduction <a id='introduction'></a>

Beta Bank customers are leaving: little by little, chipping away every month. The bankers figured out it’s cheaper to save the existing customers rather than to attract new ones.

We need to predict whether a customer will leave the bank soon. Using the provided data on clients’ past behavior and termination of contracts with the bank (`/datasets/Churn.csv`), build a model with the maximum possible F1 score. 

To pass the project, you need an F1 score of at least 0.59. Check the F1 for the test set.
Additionally, measure the AUC-ROC metric and compare it with the F1.

[Back to Contents](#back)

# Data Overview <a id='data_overview'></a>

## Initialization <a id='initialization'></a> <a class="tocSkip">

In [17]:
# Loading all the libraries
import pandas as pd
import numpy as np
from scipy.stats import randint

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split, GridSearchCV, RandomizedSearchCV, StratifiedKFold
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, confusion_matrix

## Load data <a id='load_data'></a> <a class="tocSkip">

In [2]:
# Reading the dataframe file and storing it to churn_df
churn_df = pd.read_csv('/datasets/Churn.csv')

# Prepare the data <a id='prepare_data'></a>

In [3]:
# Print the general/summary information about the DataFrame
churn_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [4]:
# Print a sample of the data
display(churn_df.head())

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

<div class="alert alert-success">
<b>Reviewer's comment</b>

The data was loaded and inspected!

</div>

## Fix Data <a id='fix_data'></a> <a class="tocSkip">

In [5]:
# the list of column names in the table
print(churn_df.columns)

Index(['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography',
       'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Exited'],
      dtype='object')


In [6]:
# renaming columns
churn_df= churn_df.rename(columns = {
    'RowNumber':'row_number',
    'CustomerId':'customer_id',
    'Surname':'surname',
    'CreditScore':'credit_score',
    'Geography':'geography',
    'Gender':'gender',
    'Age':'age',
    'Tenure':'tenure',
    'Balance':'balance',
    'NumOfProducts':'num_of_products',
    'HasCrCard':'has_cr_card',
    'IsActiveMember':'is_active_member',
    'EstimatedSalary':'estimated_salary',
    'Exited':'exited'
})

In [7]:
# checking result: the list of column names
print(churn_df.columns)

display(churn_df.head())

Index(['row_number', 'customer_id', 'surname', 'credit_score', 'geography',
       'gender', 'age', 'tenure', 'balance', 'num_of_products', 'has_cr_card',
       'is_active_member', 'estimated_salary', 'exited'],
      dtype='object')


row_number  customer_id   surname  credit_score geography  gender  age  \
0           1     15634602  Hargrave           619    France  Female   42   
1           2     15647311      Hill           608     Spain  Female   41   
2           3     15619304      Onio           502    France  Female   42   
3           4     15701354      Boni           699    France  Female   39   
4           5     15737888  Mitchell           850     Spain  Female   43   

   tenure    balance  num_of_products  has_cr_card  is_active_member  \
0     2.0       0.00                1            1                 1   
1     1.0   83807.86                1            0                 1   
2     8.0  159660.80                3            1                 0   
3     1.0       0.00                2            0                 0   
4     2.0  125510.82                1            1                 1   

   estimated_salary  exited  
0         101348.88       1  
1         112542.58       0  
2         113931.57       1  
3          93826.63       0  
4          79084.10       0

Formatting of the column names was changed to snake case for consistency. Data types for columns do not need to be changed.

<div class="alert alert-success">
<b>Reviewer's comment</b>

Alright!

</div>

## Check for Duplicates <a id='duplicates'></a> <a class="tocSkip">

In [8]:
# checking for duplicated rows
print(churn_df.duplicated().sum())
print()

# checking for duplicate row_number
print(churn_df.duplicated(subset='row_number').sum())
print()

# checking for duplicate customer_id
print(churn_df.duplicated(subset='customer_id').sum())
print()

# checking for implicit duplicates
sorted_geography = sorted(churn_df['geography'].unique())
print(sorted_geography)

0

0

0

['France', 'Germany', 'Spain']


No duplicate rows or duplicates within columns were found that would impact the analysis.

<div class="alert alert-success">
<b>Reviewer's comment</b>

Good!

</div>

## Check for Missing Values <a id='missing_values'></a> <a class="tocSkip">

In [9]:
# calculating missing values
print(churn_df.isna().sum())

row_number            0
customer_id           0
surname               0
credit_score          0
geography             0
gender                0
age                   0
tenure              909
balance               0
num_of_products       0
has_cr_card           0
is_active_member      0
estimated_salary      0
exited                0
dtype: int64


In [10]:
# Calculating missing tenure row percentages
print('% missing tenure rows:')
churn_df['tenure'].isna().mean()*100

% missing tenure rows:


9.09

In [11]:
# Find rows with missing values
rows_with_missing_values = churn_df[churn_df.isnull().any(axis=1)]

# Display the rows with missing values
display(rows_with_missing_values)

row_number  customer_id    surname  credit_score geography  gender  age  \
30            31     15589475    Azikiwe           591     Spain  Female   39   
48            49     15766205        Yin           550   Germany    Male   38   
51            52     15768193  Trevisani           585   Germany    Male   36   
53            54     15702298   Parkhill           655   Germany    Male   41   
60            61     15651280     Hunter           742   Germany    Male   35   
...          ...          ...        ...           ...       ...     ...  ...   
9944        9945     15703923    Cameron           744   Germany    Male   41   
9956        9957     15707861      Nucci           520    France  Female   46   
9964        9965     15642785    Douglas           479    France    Male   34   
9985        9986     15586914     Nepean           659    France    Male   36   
9999       10000     15628319     Walker           792    France  Female   28   

      tenure    balance  num_of_products  has_cr_card  is_active_member  \
30       NaN       0.00                3            1                 0   
48       NaN  103391.38                1            0                 1   
51       NaN  146050.97                2            0                 0   
53       NaN  125561.97                1            0                 0   
60       NaN  136857.00                1            0                 0   
...      ...        ...              ...          ...               ...   
9944     NaN  190409.34                2            1                 1   
9956     NaN   85216.61                1            1                 0   
9964     NaN  117593.48                2            0                 0   
9985     NaN  123841.49                2            1                 0   
9999     NaN  130142.79                1            1                 0   

      estimated_salary  exited  
30           140469.38       1  
48            90878.13       0  
51            86424.57       0  
53           164040.94       1  
60            84509.57       0  
...                ...     ...  
9944         138361.48       0  
9956         117369.52       1  
9964         113308.29       0  
9985          96833.00       0  
9999          38190.78       0  

[909 rows x 14 columns]

In [12]:
# Drop rows with missing "tenure" data
churn_df.dropna(subset=['tenure'], inplace=True)

# Confirming no more missing values
print(churn_df.isna().sum())

row_number          0
customer_id         0
surname             0
credit_score        0
geography           0
gender              0
age                 0
tenure              0
balance             0
num_of_products     0
has_cr_card         0
is_active_member    0
estimated_salary    0
exited              0
dtype: int64


The missing values were dropped to maintain numeric data type for creating a model. The only missing values in the dataframe are limited to only within the 'tenure' column. Due to the high number of variables affecting a customer's tenure and relatively small amount of missing values, accounting for about 9% of the tenure data, the missing values will also not be imputed to maintain the data integrity and will be left as NaN. This is acceptable since the missing tenure data will not be crucial for the analysis and is not believed to introduce bias. The missing values also seem to be missing at random with no particualr trend, also making it more acceptable to drop.

<div class="alert alert-success">
<b>Reviewer's comment</b>

Missing values were dealt with reasonably! Nice explanation of your thought process!
    

</div>

## Converting Data Types <a id='data_types'></a> <a class="tocSkip">

In [13]:
# Converting categorical variables, 'geography' and 'gender', to numerical form for use as input for a logistic regression model
categorical_columns = ['geography', 'gender']

# Apply one-hot encoding to these categorical columns
churn_df_encoded = pd.get_dummies(churn_df, columns=categorical_columns, drop_first=False)

display(churn_df_encoded)

row_number  customer_id    surname  credit_score  age  tenure  \
0              1     15634602   Hargrave           619   42     2.0   
1              2     15647311       Hill           608   41     1.0   
2              3     15619304       Onio           502   42     8.0   
3              4     15701354       Boni           699   39     1.0   
4              5     15737888   Mitchell           850   43     2.0   
...          ...          ...        ...           ...  ...     ...   
9994        9995     15719294       Wood           800   29     2.0   
9995        9996     15606229   Obijiaku           771   39     5.0   
9996        9997     15569892  Johnstone           516   35    10.0   
9997        9998     15584532        Liu           709   36     7.0   
9998        9999     15682355  Sabbatini           772   42     3.0   

        balance  num_of_products  has_cr_card  is_active_member  \
0          0.00                1            1                 1   
1      83807.86                1            0                 1   
2     159660.80                3            1                 0   
3          0.00                2            0                 0   
4     125510.82                1            1                 1   
...         ...              ...          ...               ...   
9994       0.00                2            0                 0   
9995       0.00                2            1                 0   
9996   57369.61                1            1                 1   
9997       0.00                1            0                 1   
9998   75075.31                2            1                 0   

      estimated_salary  exited  geography_France  geography_Germany  \
0            101348.88       1                 1                  0   
1            112542.58       0                 0                  0   
2            113931.57       1                 1                  0   
3             93826.63       0                 1                  0   
4             79084.10       0                 0                  0   
...                ...     ...               ...                ...   
9994         167773.55       0                 1                  0   
9995          96270.64       0                 1                  0   
9996         101699.77       0                 1                  0   
9997          42085.58       1                 1                  0   
9998          92888.52       1                 0                  1   

      geography_Spain  gender_Female  gender_Male  
0                   0              1            0  
1                   1              1            0  
2                   0              1            0  
3                   0              1            0  
4                   1              1            0  
...               ...            ...          ...  
9994                0              1            0  
9995                0              0            1  
9996                0              0            1  
9997                0              1            0  
9998                0              0            1  

[9091 rows x 17 columns]

One-Hot-Encoding was used to convert the categorical variables that contained multiple categories into numeric values.

<div class="alert alert-success">
<b>Reviewer's comment</b>

Categorical features were encoded successfully

</div>

# Class Balance Examination <a id='class_balance'></a>

In [14]:
# Count the occurrences of each class in the "exited" column
class_counts = churn_df_encoded['exited'].value_counts()

# Display the class counts
print(class_counts)

0    7237
1    1854
Name: exited, dtype: int64


In the above Series, it shows two rows: one for class 0 (customers who didn't exit) and one for class 1 (customers who exited). The results indicate that there is indeed a class imbalance, with class 0 being dominant since there are 7273 samples with class 0 and 1854 samples with class 1.

<div class="alert alert-success">
<b>Reviewer's comment</b>

Class imbalance was noted

</div>

## Model without Accounting for Imbalance <a id='raw_model'></a> <a class="tocSkip">

In [15]:
# Set a fixed random seed for reproducibility
np.random.seed(42)

# Select relevant numeric features (exclude non-numeric and non-predictive columns)
numeric_features = ['credit_score', 'age', 'tenure', 'balance', 'num_of_products', 'has_cr_card', 'is_active_member', 'estimated_salary', 'geography_Germany', 'geography_Spain', 'gender_Male']

# Create a new DataFrame with only the selected features
X_numeric = churn_df_encoded[numeric_features]

# Separate features (X) and the target variable (y)
X = X_numeric
y = churn_df_encoded['exited']

# Split your data into train, validation, and test sets
X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.3, random_state=42)
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)

# Create and train a Random Forest classifier
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

rf_classifier = RandomForestClassifier(random_state=42)
grid_search = GridSearchCV(rf_classifier, param_grid, cv=5, scoring='f1', n_jobs=-1)
grid_search.fit(X_train, y_train)

# Step 4: Model Selection and Evaluation
best_model = grid_search.best_estimator_

# Make predictions on the validation set
y_pred_val = best_model.predict(X_val)

# Evaluate the model
accuracy = accuracy_score(y_val, y_pred_val)
precision = precision_score(y_val, y_pred_val)
recall = recall_score(y_val, y_pred_val)
f1 = f1_score(y_val, y_pred_val)
roc_auc = roc_auc_score(y_val, best_model.predict_proba(X_val)[:, 1])
confusion = confusion_matrix(y_val, y_pred_val)

# Print the evaluation metrics
print(f"Accuracy: {accuracy}")
print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"F1 Score: {f1}")
print(f"AUC-ROC Score: {roc_auc}")
print("Confusion Matrix:")
print(confusion)

# Finally, evaluate the model on the test set (unseen data)
y_pred_test = best_model.predict(X_test)
test_accuracy = accuracy_score(y_test, y_pred_test)
print(f"Test Accuracy: {test_accuracy}")


Accuracy: 0.8563049853372434
Precision: 0.7894736842105263
Recall: 0.4225352112676056
F1 Score: 0.5504587155963302
AUC-ROC Score: 0.8642703442879498
Confusion Matrix:
[[1048   32]
 [ 164  120]]
Test Accuracy: 0.8629032258064516


<div class="alert alert-danger">
<s><b>Reviewer's comment</b>

1. The data was split into train and test sets. Later you split the same data into train and validation. Unfortunately that doesn't make much sense: we need to split the data into three parts from the start. The train set is then used to train the models, validation set is used to compare different models and tune their hyperparameters and the test set is used at the very end, to evaluate the final model for an unbiased estimate of its generalization performance. Although, as you're using cross-validation, it is possible to use just two sets: train and test, with train set used for cross-validation (you can compare different models using cross-validation as well as tune hyperpameteres), and the test set used to evaluate the final model.

2. To calculate ROC-AUC, we need different inputs from the other metrics: instead of binary predictions (method `predict`) we need 'probabilities' (method `predict_proba`). The reason is that the ROC curve is constructed by varying the threshold of assigning positive class between 0 and 1, and for binary predictions the threshold is predefined

</div>

<div class="alert alert-warning">
<b>Reviewer's comment V2</b>

Both problems were fixed!
    
Although, as you're using cross-validation, we don't really need to have both validation and test, we can just use cross-validation for hyperparameter tuning and model selection, and then evaluate the final model on the test set

</div>

The trained model, without accounting for the class imbalance, performed relatively well.

The model achieved an accuracy of approximately 85.6% and precision of approximately 78.9%, meaning that when the model predicts a customer will churn, it is correct about 78.9% of the time.

The recall of approximately 42.2% indicates that the model correctly identifies about 42.2% of all actual churn cases. Recall measures the proportion of actual positive cases that the model correctly predicted as positive.

The F1 score of approximately 55.0% is the harmonic mean of precision and recall. It provides a balance between precision and recall, with higher values indicating a better balance.

The AUC-ROC score of approximately 86.4% is a measure of the model's ability to distinguish between positive and negative cases. An AUC-ROC score of 0.5 indicates that a model's performance is similar to random guessing. Since the AUC-ROC score is above 0.5, it performs better than random guessing.

The confusion matrix provides a breakdown of the model's predictions. In this case, the model made 120 true positives, 1048 true negatives, 32 false positives, and 164 false negatives..

This trained model without accounting for the class imbalance was made for comparison purposes. Improvements in model performance will be necessary to meet the objectives of predicting churn with a higher F1 score (at least 0.59), as well as addressing the class imbalance, though it was not too far off in achieving the required score.

<div class="alert alert-info">
  Makes sense for comparison purposes to use the same type of model for balanced and not balanced. instead of doing an additional two models later for random forest and decision tree without balancing, I decided to just show the unbalanced as a random forest model instead of logistic regression since I was pretty certain to begin with that I'd end up using the random forest model anyways
</div>

<div class="alert alert-success">
<b>Reviewer's comment V2</b>

Ok, great!

</div>

<div class="alert alert-warning">
<b>Reviewer's comment</b>

Alright, you trained a logistic regression model without taking the imbalance into account first. But it's difficult to see the effect of balancing because you're training different models in the next section. It would be nice if you tried those models without balancing as well, or if you tried applying balancing to the logistic regression model

</div>

## Fixing Class Imbalance <a id='fixing_class_imbalance'></a> <a class="tocSkip">

### Upsampling Random Forest Model <a id='upsampling_random_forest'></a> <a class="tocSkip">

In [18]:
# Set a fixed random seed for reproducibility
np.random.seed(42)

# Separate features (X) and the target variable (y)
X = X_numeric
y = churn_df_encoded['exited']

# Step 1: Split your data into train (60%), validation (20%), and test (20%) sets
X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.4, random_state=42)
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)

# Step 2: Define a StratifiedKFold cross-validator with 5 folds
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# Step 3: Hyperparameter Tuning (as previously described)
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

rf_classifier = RandomForestClassifier(random_state=42)

# Perform grid search with cross-validation
best_f1_score = -1  # Initialize with a low value
best_model = None

for train_index, val_index in cv.split(X_train, y_train):
    X_train_fold, X_val_fold = X_train.iloc[train_index], X_train.iloc[val_index]
    y_train_fold, y_val_fold = y_train.iloc[train_index], y_train.iloc[val_index]
    
    # Step 4: Apply Oversampling within each fold
    majority_indices = np.where(y_train_fold == 0)[0]
    minority_indices = np.where(y_train_fold == 1)[0]
    
    # Calculate the oversampling ratio
    oversampling_ratio = len(majority_indices) // len(minority_indices)
    
    # Randomly oversample the minority class within the fold
    oversampled_minority_indices = np.random.choice(minority_indices, size=len(minority_indices) * oversampling_ratio, replace=True)
    
    # Combine the oversampled minority and majority class
    oversampled_indices = np.concatenate((majority_indices, oversampled_minority_indices))
    
    # Create the oversampled training set for this fold
    X_train_fold = X_train_fold.iloc[oversampled_indices]
    y_train_fold = y_train_fold.iloc[oversampled_indices]
    
    rf_classifier.fit(X_train_fold, y_train_fold)
    y_pred_val_fold = rf_classifier.predict(X_val_fold)
    f1_val_fold = f1_score(y_val_fold, y_pred_val_fold)
    
    if f1_val_fold > best_f1_score:
        best_f1_score = f1_val_fold
        best_model = rf_classifier

# Step 5: Model Selection and Evaluation (including AUC-ROC) on the validation set
y_pred_val = best_model.predict(X_val)
f1_val = f1_score(y_val, y_pred_val)

# Calculate AUC-ROC on the validation set
y_prob_val = best_model.predict_proba(X_val)[:, 1]
roc_auc_val = roc_auc_score(y_val, y_prob_val)

print("Results after Oversampling with Cross-Validation:")
print(f"Best Parameters: {best_model.get_params()}")
print(f"F1 Score on Validation Set: {f1_val}")
print(f"AUC-ROC Score on Validation Set: {roc_auc_val}")

# Step 6: Evaluate the final model on the test set
y_pred_test = best_model.predict(X_test)
f1_test = f1_score(y_test, y_pred_test)

# Calculate AUC-ROC on the test set
y_prob_test = best_model.predict_proba(X_test)[:, 1]
roc_auc_test = roc_auc_score(y_test, y_prob_test)

print("Results on Test Set:")
print(f"F1 Score on Test Set: {f1_test}")
print(f"AUC-ROC Score on Test Set: {roc_auc_test}")

Results after Oversampling with Cross-Validation:
Best Parameters: {'bootstrap': True, 'ccp_alpha': 0.0, 'class_weight': None, 'criterion': 'gini', 'max_depth': None, 'max_features': 'auto', 'max_leaf_nodes': None, 'max_samples': None, 'min_impurity_decrease': 0.0, 'min_impurity_split': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'min_weight_fraction_leaf': 0.0, 'n_estimators': 100, 'n_jobs': None, 'oob_score': False, 'random_state': 42, 'verbose': 0, 'warm_start': False}
F1 Score on Validation Set: 0.5419354838709677
AUC-ROC Score on Validation Set: 0.8439708345433856
Results on Test Set:
F1 Score on Test Set: 0.5790349417637272
AUC-ROC Score on Test Set: 0.8488455427598525


<div class="alert alert-info">
  I did not use imblearn because even though I am working through my project on the tripleten platform, it seems to not support the imblearn module as I would get this error 
    
    ModuleNotFoundError                       Traceback (most recent call last)
/tmp/ipykernel_27/1036495727.py in <module>
      3 from sklearn.model_selection import train_test_split, GridSearchCV, StratifiedKFold
      4 from sklearn.metrics import f1_score, roc_auc_score
----> 5 from imblearn.over_sampling import RandomOverSampler
      6 from imblearn.pipeline import Pipeline
      7 from sklearn.ensemble import RandomForestClassifier

ModuleNotFoundError: No module named 'imblearn'
</div>

<div class="alert alert-warning">
<b>Reviewer's comment V2</b>

Indeed, it is not installed by default, but you can install it like this:
    
```python
!pip install --user imblearn 
```
    
Note that after installing it, the kernel needs to be reloaded before it can see the new library.
    
    
In any case, it's cool that you tried to implement it on your own, as it allows you to understand how it all works under the hood.
    
There are a couple of issues though:
    
1. First of all, in k-fold cross-validation, we want to train k models of the same kind and average their scores, it doesn't make sense to report the best fold's score
    
2. And second, in your code in this cell, there is actually no hypeprarameter tuning: only the model with the default hyperparameters is trained. You'd need to add an outer loop wrapping the cross-validation to do that. Roughly it would look like this (this is pseudocode just for understanding):
    
```python
for hyperparameter_values in hyperparameter_values_list:
    f1_scores_folds = []
    for train_index, val_index in cv.split(X_train, y_train):
        ...
        model = RandomForestClassifier(...) # you need to set hypeprarameters here in some way depending on how you store them
        model.fit(X_train_fold, y_train_fold)
        f1_val_fold = f1_score(y_val_fold, y_pred_val_fold)
        f1_scores_folds.append(f1_val_fold)
    f1_score_ = np.mean(f1_scores_folds)
    if f1_score_ > best_f1_score:
        best_f1_score = f1_score_
        best_model = model
    
```
    
This is basically what GridSearchCV does under the hood.

</div>

In [2]:
import numpy as np
np.mean([1,2,3])

2.0

<div class="alert alert-warning">
<b>Reviewer's comment</b>
    
It's cool that you used a grid search with cross-validation to tune hyperparameters, but it's not really appropriate to use oversampled data for cross-validation. The correct process is to apply oversampling in each fold separately, e.g. using [imblearn pipelines](https://imbalanced-learn.org/stable/references/generated/imblearn.pipeline.Pipeline.html#imblearn.pipeline.Pipeline) and [oversamplers](https://imbalanced-learn.org/stable/references/over_sampling.html)

</div>

In summary, the model's performance after applying oversampling with cross-validation resulted in an F1 score of approximately 0.5420 on the validation set and approximately 0.5790 on the test set. The AUC-ROC score was approximately 0.8440 on the validation set and approximately 0.8488 on the test set. These results for the oversampling approach did not significantly differ from the model's performance compared to the previous results without oversampling to balance the data, which had an F1 score of 0.55 and AUC-ROC score of 0.86. Additionally, the oversampling approach did not achieve the required F1 score of 0.59 on the validation set as well.

### Downsampling Random Forest <a id='downsampling_random_forest'></a> <a class="tocSkip">

In [19]:
# Set a fixed random seed for reproducibility
np.random.seed(42)

# Separate features (X) and the target variable (y)
X = X_numeric
y = churn_df_encoded['exited']

# Step 1: Split your data into train (60%), validation (20%), and test (20%) sets
X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.4, random_state=42)
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)

# Step 2: Define a StratifiedKFold cross-validator with 5 folds
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# Step 3: Hyperparameter Tuning (as previously described)
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

rf_classifier = RandomForestClassifier(random_state=42)

# Perform grid search with cross-validation
best_f1_score = -1  # Initialize with a low value
best_model = None

for train_index, val_index in cv.split(X_train, y_train):
    X_train_fold, X_val_fold = X_train.iloc[train_index], X_train.iloc[val_index]
    y_train_fold, y_val_fold = y_train.iloc[train_index], y_train.iloc[val_index]
    
    # Step 4: Apply Downsampling within each fold
    majority_indices = np.where(y_train_fold == 0)[0]
    minority_indices = np.where(y_train_fold == 1)[0]
    
    # Randomly undersample the majority class within the fold
    undersampled_majority_indices = np.random.choice(majority_indices, size=len(minority_indices), replace=False)
    
    # Combine the undersampled majority and minority class
    undersampled_indices = np.concatenate((undersampled_majority_indices, minority_indices))
    
    # Create the undersampled training set for this fold
    X_train_fold = X_train_fold.iloc[undersampled_indices]
    y_train_fold = y_train_fold.iloc[undersampled_indices]
    
    rf_classifier.fit(X_train_fold, y_train_fold)
    y_pred_val_fold = rf_classifier.predict(X_val_fold)
    f1_val_fold = f1_score(y_val_fold, y_pred_val_fold)
    
    if f1_val_fold > best_f1_score:
        best_f1_score = f1_val_fold
        best_model = rf_classifier

# Step 5: Model Selection and Evaluation (including AUC-ROC) on the validation set
y_pred_val = best_model.predict(X_val)
f1_val = f1_score(y_val, y_pred_val)

# Calculate AUC-ROC on the validation set
y_prob_val = best_model.predict_proba(X_val)[:, 1]
roc_auc_val = roc_auc_score(y_val, y_prob_val)

print("Results after Downsampling with Cross-Validation:")
print(f"Best Parameters: {best_model.get_params()}")
print(f"F1 Score on Validation Set: {f1_val}")
print(f"AUC-ROC Score on Validation Set: {roc_auc_val}")

# Step 6: Evaluate the final model on the test set
y_pred_test = best_model.predict(X_test)
f1_test = f1_score(y_test, y_pred_test)

# Calculate AUC-ROC on the test set
y_prob_test = best_model.predict_proba(X_test)[:, 1]
roc_auc_test = roc_auc_score(y_test, y_prob_test)

print("Results on Test Set:")
print(f"F1 Score on Test Set: {f1_test}")
print(f"AUC-ROC Score on Test Set: {roc_auc_test}")

Results after Downsampling with Cross-Validation:
Best Parameters: {'bootstrap': True, 'ccp_alpha': 0.0, 'class_weight': None, 'criterion': 'gini', 'max_depth': None, 'max_features': 'auto', 'max_leaf_nodes': None, 'max_samples': None, 'min_impurity_decrease': 0.0, 'min_impurity_split': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'min_weight_fraction_leaf': 0.0, 'n_estimators': 100, 'n_jobs': None, 'oob_score': False, 'random_state': 42, 'verbose': 0, 'warm_start': False}
F1 Score on Validation Set: 0.5991649269311066
AUC-ROC Score on Validation Set: 0.8476247358433014
Results on Test Set:
F1 Score on Test Set: 0.5739320920043811
AUC-ROC Score on Test Set: 0.8473609968261432


<div class="alert alert-warning">
<b>Reviewer's comment</b>

Same as for upsampling, the correct way to apply downsampling in cross-validation is to downsample the train subset in each fold separately

</div>

In summary, the model's performance after applying downsampling with cross-validation resulted in an F1 score of approximately 0.5992 on the validation set and approximately 0.5739 on the test set. The AUC-ROC score was approximately 0.8476 on the validation set and approximately 0.8474 on the test set.

These results indicate that the downsampling approach with cross-validation has improved the model's F1 score on the validation set compared to the previous results that used an oversampling approach. However, the F1 score on the test set is slightly lower than the validation set. The AUC-ROC scores suggest good discrimination ability between classes, and the model performs well in terms of class separation. Additionally, this model with the downsampling approach was able to meet the required F1 score for the validation set of at least 0.59.

### Upsampling Decision Tree Model <a id='upsampling_decision_tree'></a> <a class="tocSkip">

In [20]:
# Set a fixed random seed for reproducibility
np.random.seed(42)

# Split your data into training and validation sets (adjust the test_size as needed)
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# Step 1: Identify indices for the majority (non-churn) and minority (churn) classes
majority_indices = np.where(y_train == 0)[0]
minority_indices = np.where(y_train == 1)[0]

# Step 2: Randomly oversample the minority class
oversampled_minority_indices = np.random.choice(minority_indices, size=len(majority_indices), replace=True)

# Combine the majority and oversampled minority class
oversampled_indices = np.concatenate((majority_indices, oversampled_minority_indices))

# Create the oversampled training set
oversampled_X = X_train.iloc[oversampled_indices]
oversampled_y = y_train.iloc[oversampled_indices]

# Step 3: Hyperparameter Tuning for Decision Tree
param_grid = {
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

decision_tree_classifier = DecisionTreeClassifier(random_state=42)
grid_search = GridSearchCV(decision_tree_classifier, param_grid, cv=5, scoring='f1', n_jobs=-1)
grid_search.fit(oversampled_X, oversampled_y)

# Step 4: Model Selection and Evaluation (including AUC-ROC)
best_model_upsampled = grid_search.best_estimator_
y_pred_val_upsampled = best_model_upsampled.predict(X_val)
f1_val_upsampled = f1_score(y_val, y_pred_val_upsampled)

# Calculate AUC-ROC on the validation set
y_prob_val_upsampled = best_model_upsampled.predict_proba(X_val)[:, 1]
roc_auc_val_upsampled = roc_auc_score(y_val, y_prob_val_upsampled)

print("Results for Upsampling with Decision Tree Classifier:")
print(f"Best Parameters: {grid_search.best_params_}")
print(f"F1 Score on Validation Set: {f1_val_upsampled}")
print(f"AUC-ROC Score on Validation Set: {roc_auc_val_upsampled}")

Results for Upsampling with Decision Tree Classifier:
Best Parameters: {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2}
F1 Score on Validation Set: 0.4885290148448043
AUC-ROC Score on Validation Set: 0.677618748033973


The F1 score is approximately 0.49 and the AUC-ROC score is approximately 0.68. Overall, the upsampling approach with a Decision Tree Classifier resulted in a moderate F1 score and a decent AUC-ROC score. While the F1 score could be improved, the model demonstrates the ability to distinguish between churn and non-churn customers to some extent. However, this model does not pass in achieving the required F1 score of 0.59.

### Downsampling Decision Tree Model <a id='downsampling_decision_tree'></a> <a class="tocSkip">

In [21]:
# Set a fixed random seed for reproducibility
np.random.seed(42)

# Split your data into training and validation sets (adjust the test_size as needed)
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# Step 1: Identify indices for the majority (non-churn) and minority (churn) classes
majority_indices = np.where(y_train == 0)[0]
minority_indices = np.where(y_train == 1)[0]

# Step 2: Randomly undersample the majority class
undersampled_majority_indices = np.random.choice(majority_indices, size=len(minority_indices), replace=False)

# Combine the undersampled majority and minority class
undersampled_indices = np.concatenate((undersampled_majority_indices, minority_indices))

# Create the undersampled training set
undersampled_X = X_train.iloc[undersampled_indices]
undersampled_y = y_train.iloc[undersampled_indices]

# Step 3: Hyperparameter Tuning for Decision Tree
param_grid = {
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

decision_tree_classifier = DecisionTreeClassifier(random_state=42)
grid_search = GridSearchCV(decision_tree_classifier, param_grid, cv=5, scoring='f1', n_jobs=-1)
grid_search.fit(undersampled_X, undersampled_y)

# Step 4: Model Selection and Evaluation (including AUC-ROC)
best_model_downsampled = grid_search.best_estimator_
y_pred_val_downsampled = best_model_downsampled.predict(X_val)
f1_val_downsampled = f1_score(y_val, y_pred_val_downsampled)

# Calculate AUC-ROC on the validation set
y_prob_val_downsampled = best_model_downsampled.predict_proba(X_val)[:, 1]
roc_auc_val_downsampled = roc_auc_score(y_val, y_prob_val_downsampled)

print("Results for Downsampling with Decision Tree Classifier:")
print(f"Best Parameters: {grid_search.best_params_}")
print(f"F1 Score on Validation Set: {f1_val_downsampled}")
print(f"AUC-ROC Score on Validation Set: {roc_auc_val_downsampled}")

Results for Downsampling with Decision Tree Classifier:
Best Parameters: {'max_depth': 10, 'min_samples_leaf': 4, 'min_samples_split': 10}
F1 Score on Validation Set: 0.5137614678899083
AUC-ROC Score on Validation Set: 0.7761671261773033


The F1 score is approximately 0.51 and the AUC-ROC score is approximately 0.78. Overall, the downsampling approach with a Decision Tree Classifier was slightly improved compared to the previous upsampling approach. However, this model still does not pass in achieving the required F1 score of 0.59.

<div class="alert alert-success">
<b>Reviewer's comment</b>

Great, you successfully applied two different balancing methods and trained a couple of different models with upsampled/donwsampled data

</div>

## Final Model <a id='final_model'></a> <a class="tocSkip">

In [22]:
# Set a fixed random seed for reproducibility
np.random.seed(42)

# Separate features (X) and the target variable (y)
X = X_numeric
y = churn_df_encoded['exited']

# Step 1: Split your data into train (60%), validation (20%), and test (20%) sets
X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.4, random_state=42)
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)

# Step 2: Define a StratifiedKFold cross-validator with 5 folds
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# Step 3: Hyperparameter Tuning (as previously described)
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

rf_classifier = RandomForestClassifier(random_state=42)

# Perform grid search with cross-validation
best_f1_score = -1  # Initialize with a low value
best_model = None

for train_index, val_index in cv.split(X_train, y_train):
    X_train_fold, X_val_fold = X_train.iloc[train_index], X_train.iloc[val_index]
    y_train_fold, y_val_fold = y_train.iloc[train_index], y_train.iloc[val_index]
    
    # Step 4: Apply Downsampling within each fold
    majority_indices = np.where(y_train_fold == 0)[0]
    minority_indices = np.where(y_train_fold == 1)[0]
    
    # Randomly undersample the majority class within the fold
    undersampled_majority_indices = np.random.choice(majority_indices, size=len(minority_indices), replace=False)
    
    # Combine the undersampled majority and minority class
    undersampled_indices = np.concatenate((undersampled_majority_indices, minority_indices))
    
    # Create the undersampled training set for this fold
    X_train_fold = X_train_fold.iloc[undersampled_indices]
    y_train_fold = y_train_fold.iloc[undersampled_indices]
    
    rf_classifier.fit(X_train_fold, y_train_fold)
    y_pred_val_fold = rf_classifier.predict(X_val_fold)
    f1_val_fold = f1_score(y_val_fold, y_pred_val_fold)
    
    if f1_val_fold > best_f1_score:
        best_f1_score = f1_val_fold
        best_model = rf_classifier

# Step 5: Model Selection and Evaluation (including AUC-ROC) on the validation set
y_pred_val = best_model.predict(X_val)
f1_val = f1_score(y_val, y_pred_val)

# Calculate AUC-ROC on the validation set
y_prob_val = best_model.predict_proba(X_val)[:, 1]
roc_auc_val = roc_auc_score(y_val, y_prob_val)

print("Results after Downsampling with Cross-Validation:")
print(f"Best Parameters: {best_model.get_params()}")
print(f"F1 Score on Validation Set: {f1_val}")
print(f"AUC-ROC Score on Validation Set: {roc_auc_val}")

# Step 6: Evaluate the final model on the test set
y_pred_test = best_model.predict(X_test)
f1_test = f1_score(y_test, y_pred_test)

# Calculate AUC-ROC on the test set
y_prob_test = best_model.predict_proba(X_test)[:, 1]
roc_auc_test = roc_auc_score(y_test, y_prob_test)

print("Results on Test Set:")
print(f"F1 Score on Test Set: {f1_test}")
print(f"AUC-ROC Score on Test Set: {roc_auc_test}")

Results after Downsampling with Cross-Validation:
Best Parameters: {'bootstrap': True, 'ccp_alpha': 0.0, 'class_weight': None, 'criterion': 'gini', 'max_depth': None, 'max_features': 'auto', 'max_leaf_nodes': None, 'max_samples': None, 'min_impurity_decrease': 0.0, 'min_impurity_split': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'min_weight_fraction_leaf': 0.0, 'n_estimators': 100, 'n_jobs': None, 'oob_score': False, 'random_state': 42, 'verbose': 0, 'warm_start': False}
F1 Score on Validation Set: 0.5991649269311066
AUC-ROC Score on Validation Set: 0.8476247358433014
Results on Test Set:
F1 Score on Test Set: 0.5739320920043811
AUC-ROC Score on Test Set: 0.8473609968261432


The final model chosen was the Random Forest model using the downsampling approach since it was able to achieve the required F1 score of at least 0.59. When a specific performance metric such as an F1 score is set as a requirement, you typically aim to achieve that performance on the validation set, not the test set. The F1 score on the validation set passed with a score of 0.599 with an AUC-ROC score is approximately 0.85 on the validation set. These results suggest that the model performs well in identifying churn and non-churn customers in the dataset.

<div class="alert alert-success">
<b>Reviewer's comment</b>

Awesome!

</div>

# Conclusion <a id='conclusion'></a>

The original dataset was found to have a class imbalance, favoring class 0 (customers who did not exit). For comparison purposes, a model was trained without taking into account the imbalance. The model actually performed relatively well, having a  F1 score of 0.55, indicating decent performance of a classification model and an AUC-ROC score of 0.86. An AUC-ROC score of 0.5 indicates random chance, while a higher score suggests better model performance, thus the model's performance was better than random guessing.

Both upsampling and downsampling approaches were used to account for the class imbalance, trying both the Random Forest and Decision Tree model. The upsampling approach for the Random Forest model achieved a reasonably good F1 score of 0.54 and a strong AUC-ROC score of 0.84 on the validation set. However, it did not achieve the required F1 score to pass the project. The downsampling approach for the Random Forest model was able to achieve the passing F1 score of at least 0.59 and an AUC-ROC score of 0.85 on the validation set.

The Decision Tree model was also tested for comparison purposes, but for both the upsampling and downsampling approach, while it had a moderate F1 score of approximately 0.5 for and decent AUC-ROC scores of approximately 0.7 for both approaches, it ultimately did not pass the required F1 score.

Based on these results, the final model chosen was the downsampling approach for the Random Forest model since it achieved a passing F1 score of 0.59 and a high AUC-ROC score of 0.85. These results suggest that the model performs well in identifying churn and non-churn customers in the dataset and performs better than random guessing since the AUC-ROC score is higher than 0.5.

<div class="alert alert-success">
<b>Reviewer's comment</b>

Excellent summary!

</div>